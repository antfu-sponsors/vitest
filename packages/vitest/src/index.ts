// TODO: deprecate <reference types="vitest" /> in favor of `<reference types="vitest/config" />`
import './node/types/vite'
import './types/global'

export {
  suite,
  test,
  describe,
  it,
  beforeAll,
  beforeEach,
  afterAll,
  afterEach,
  onTestFailed,
  onTestFinished,
} from '@vitest/runner'
export { bench } from './runtime/benchmark'
export { expectTypeOf } from './typecheck/expectTypeOf'
export { assertType } from './typecheck/assertType'

export { runOnce, isFirstRun } from './integrations/run-once'
export { createExpect, assert, should, chai, expect } from './integrations/chai'
export { vi, vitest } from './integrations/vi'
export { getRunningMode, isWatchMode } from './integrations/utils'
export { inject } from './integrations/inject'

export type { VitestUtils } from './integrations/vi'

export type { ExpectTypeOf } from './typecheck/expectTypeOf'
export type { AssertType } from './typecheck/assertType'
export type {
  /** @deprecated import from `vitest/node` instead */
  RawErrsMap,
  /** @deprecated import from `vitest/node` instead */
  TscErrorInfo,
  /** @deprecated import from `vitest/node` instead */
  CollectLineNumbers,
  /** @deprecated import from `vitest/node` instead */
  CollectLines,
  /** @deprecated import from `vitest/node` instead */
  RootAndTarget,
  /** @deprecated import from `vitest/node` instead */
  Context,
} from './typecheck/types'

export type {
  RunMode,
  TaskState,
  TaskBase,
  TaskResult,
  TaskResultPack,
  Suite,
  File,
  Test,
  Task,
  DoneCallback,
  TestFunction,
  TestOptions,
  TestAPI,
  SuiteAPI,
  HookListener,
  HookCleanupCallback,
  SuiteHooks,
  SuiteCollector,
  SuiteFactory,
  RuntimeContext,
  TestContext,
  TaskContext,
  ExtendedContext,
  Custom,
  TaskCustomOptions,
  OnTestFailedHandler,
  TaskMeta,
} from '@vitest/runner'
export type {
  RuntimeRPC,
  RunnerRPC,
} from './types/rpc'
export type {
  SnapshotData,
  SnapshotUpdateState,
  SnapshotStateOptions,
  SnapshotMatchOptions,
  SnapshotResult,
  UncheckedSnapshot,
  SnapshotSummary,
  SnapshotSerializer,
} from '@vitest/snapshot'

export type {
  ResolveIdFunction,
  WorkerRPC,
  WorkerGlobalState,
  ContextTestEnvironment,
  ContextRPC,
} from './types/worker'
export type {
  /** @deprecated import from `vitest/node` instead */
  WorkerContext,
} from './node/types/worker'

export type {
  ErrorWithDiff,
  ParsedStack,
  /** @deprecated do not use, internal helper */
  Awaitable,
  /** @deprecated do not use, internal helper */
  Nullable,
  /** @deprecated do not use, internal helper */
  Arrayable,
  /** @deprecated do not use, internal helper */
  ArgumentsType,
  /** @deprecated do not use, internal helper */
  MutableArray,
  /** @deprecated do not use, internal helper */
  Constructable,
  ModuleCache,
  UserConsoleLog,
  ModuleGraphData,
  /** @deprecated import from `vitest/node` instead */
  OnServerRestartHandler,
  ProvidedContext,
  AfterSuiteRunMeta,
} from './types/general'
export type {
  /** @deprecated import from `vitest/environments` instead */
  EnvironmentReturn,
  /** @deprecated import from `vitest/environments` instead */
  VmEnvironmentReturn,
  /** @deprecated import from `vitest/environments` instead */
  Environment,
  /** @deprecated do not use it */
  ResolvedTestEnvironment,
} from './types/environment'
export type {
  /** @deprecated import from `vitest/node` instead */
  CoverageProvider,
  /** @deprecated import from `vitest/node` instead */
  ReportContext,
  /** @deprecated import from `vitest/node` instead */
  CoverageProviderModule,
  /** @deprecated import from `vitest/node` instead */
  CoverageReporter,
  /** @deprecated import from `vitest/node` instead */
  CoverageOptions,
  /** @deprecated import from `vitest/node` instead */
  ResolvedCoverageOptions,
  /** @deprecated import from `vitest/node` instead */
  BaseCoverageOptions,
  /** @deprecated import from `vitest/node` instead */
  CoverageIstanbulOptions,
  /** @deprecated import from `vitest/node` instead */
  CoverageV8Options,
  /** @deprecated import from `vitest/node` instead */
  CustomProviderOptions,
} from './node/types/coverage'

export type { CancelReason } from '@vitest/runner'
export type { DiffOptions } from '@vitest/utils/diff'
export type {
  MockedFunction,
  MockedObject,
  MockInstance,
  Mock,
  MockContext,
  Mocked,
  MockedClass,
} from './integrations/spy'
export type { BrowserUI } from './types/ui'

export type {
  /** @deprecated import from `vitest/node` instead */
  Reporter,
} from './node/types/reporter'
export type {
  /** @deprecated import from `vitest/node` instead */
  Vitest,
} from './node/core'

export type {
  ExpectStatic,
  AsymmetricMatchersContaining,
  JestAssertion,
  Assertion,
  ExpectPollOptions,
} from '@vitest/expect'

export type {
  SerializedConfig,
  RuntimeConfig,
  SerializedCoverageConfig,
} from './runtime/config'
export type {
  /** @deprecated import from `vitest/node` instead */
  BrowserScript,
  /** @deprecated import from `vitest/node` instead */
  BrowserConfigOptions,
  /** @deprecated import from `vitest/node` instead */
  SequenceHooks,
  /** @deprecated import from `vitest/node` instead */
  SequenceSetupFiles,
  /** @deprecated import from `vitest/node` instead */
  BuiltinEnvironment,
  /** @deprecated import from `vitest/node` instead */
  VitestEnvironment,
  /** @deprecated import from `vitest/node` instead */
  Pool,
  /** @deprecated import from `vitest/node` instead */
  PoolOptions,
  /** @deprecated import from `vitest/node` instead */
  CSSModuleScopeStrategy,
  /** @deprecated import from `vitest/node` instead */
  ApiConfig,
  /** @deprecated import from `vitest/node` instead */
  JSDOMOptions,
  /** @deprecated import from `vitest/node` instead */
  HappyDOMOptions,
  /** @deprecated import from `vitest/node` instead */
  EnvironmentOptions,
  /** @deprecated import from `vitest/node` instead */
  VitestRunMode,
  /** @deprecated import from `vitest/node` instead */
  DepsOptimizationOptions,
  /** @deprecated import from `vitest/node` instead */
  TransformModePatterns,
  /** @deprecated import from `vitest/node` instead */
  InlineConfig,
  /** @deprecated import from `vitest/node` instead */
  TypecheckConfig,
  /** @deprecated import from `vitest/node` instead */
  UserConfig,
  /** @deprecated import from `vitest/node` instead */
  ResolvedConfig,
  /** @deprecated import from `vitest/node` instead */
  ProjectConfig,
  /** @deprecated import from `vitest/node` instead */
  UserWorkspaceConfig,
} from './node/types/config'

export type {
  Benchmark,
  BenchmarkResult,
  BenchFunction,
  BenchmarkAPI,
  BenchTaskResult,
  BenchOptions,
  BenchFactory,
  BenchTask,
} from './runtime/types/benchmark'
export type {
  /** @deprecated import from `vitest/node` instead */
  BenchmarkUserOptions,
} from './node/types/benchmark'

export type {
  TransformResultWithSource,
  WebSocketHandlers,
  WebSocketEvents,
  WebSocketRPC,
} from './api/types'
