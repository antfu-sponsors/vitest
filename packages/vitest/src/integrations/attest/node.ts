import type { TestProject } from '../../node/project'
import { execFileSync } from 'node:child_process'
import { mkdirSync } from 'node:fs'

async function setup() {
  // TODO: programatic API
  mkdirSync('.attest/assertions', { recursive: true })
  execFileSync('attest', ['precache', '.attest/assertions/typescript.json'], {
    stdio: 'inherit',
    env: {
      ...process.env,
      ATTEST_attestAliases: JSON.stringify(['expect']),
    },
  })
}

export async function globalSetupAttest(project: TestProject) {
  // TODO: vitest level config
  // TODO: workspace support?
  if (!process.env.VITEST_ATTEST) {
    return
  }

  await setup()
  project.onTestsRerun(() => setup())
}
