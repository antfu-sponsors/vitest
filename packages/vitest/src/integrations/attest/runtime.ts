import type { ChaiPlugin } from '@vitest/expect'
import {
  type Plugin as PrettyFormatPlugin,
  PrettyFormatSkipSnapshotError,
} from '@vitest/pretty-format'
import { addSerializer, stripSnapshotIndentation } from '@vitest/snapshot'
import { parseStacktrace } from '@vitest/utils/source-map'
import * as chai from 'chai'
import { getSnapshotClient, getTestNames } from '../snapshot/chai'

// lazy load '@ark/attest` only when enabled
let lib: typeof import('@ark/attest')
let enabled = false

const plugin: ChaiPlugin = (chai, utils) => {
  function getTypeAssertions(ctx: object) {
    const parsed = parseStacktrace(utils.flag(ctx, '__vitest_expect_stack'))
    const location = parsed[0]
    const types = lib.getTypeAssertionsAtPosition({
      file: location.file,
      method: location.method,
      line: location.line,
      char: location.column,
    })
    return types
  }

  utils.addMethod(
    chai.Assertion.prototype,
    'toMatchTypeSnapshot',
    function (
      this: Record<string, unknown>,
      message?: string,
    ) {
      let value: any
      if (enabled) {
        const types = getTypeAssertions(this)
        value = types[0][1].args[0].type
      }
      utils.flag(this, '_name', 'toMatchTypeSnapshot')
      const test = utils.flag(this, 'vitest-test')
      getSnapshotClient().assert({
        received: new AttestSnapshotWrapper(value),
        message,
        error: utils.flag(this, 'error'),
        errorMessage: utils.flag(this, 'message'),
        ...getTestNames(test),
      })
    },
  )
  utils.addMethod(
    chai.Assertion.prototype,
    'toMatchTypeInlineSnapshot',
    function __INLINE_SNAPSHOT__(
      this: Record<string, unknown>,
      inlineSnapshot?: string,
      message?: string,
    ) {
      let value: any
      if (enabled) {
        const types = getTypeAssertions(this)
        value = types[0][1].args[0].type
      }
      utils.flag(this, '_name', 'toMatchTypeInlineSnapshot')
      const isNot = utils.flag(this, 'negate')
      if (isNot) {
        throw new Error('toMatchTypeInlineSnapshot cannot be used with "not"')
      }
      const test = utils.flag(this, 'vitest-test')
      const isInsideEach = test && (test.each || test.suite?.each)
      if (isInsideEach) {
        throw new Error(
          'InlineSnapshot cannot be used inside of test.each or describe.each',
        )
      }
      if (inlineSnapshot) {
        inlineSnapshot = stripSnapshotIndentation(inlineSnapshot)
      }
      getSnapshotClient().assert({
        received: new AttestSnapshotWrapper(value),
        message,
        error: utils.flag(this, 'error'),
        errorMessage: utils.flag(this, 'message'),
        isInline: true,
        inlineSnapshot,
        ...getTestNames(test),
      })
    },
  )
}

class AttestSnapshotWrapper {
  constructor(public value?: unknown) {}
}

const prettyFormatPlugin: PrettyFormatPlugin = {
  test(val: unknown) {
    return !!(val && val instanceof AttestSnapshotWrapper)
  },
  serialize(val: AttestSnapshotWrapper, config, indentation, depth, refs, printer) {
    if (!enabled) {
      throw new PrettyFormatSkipSnapshotError()
    }
    if (typeof val.value === 'string') {
      return val.value
    }
    return printer(
      val.value,
      config,
      indentation,
      depth,
      refs,
    )
  },
}

export async function setupAttest() {
  chai.use(plugin)
  addSerializer(prettyFormatPlugin)
  // TODO: vitest config
  enabled = !!globalThis.process?.env.VITEST_ATTEST
  if (enabled) {
    lib = await import('@ark/attest')
  }
}
