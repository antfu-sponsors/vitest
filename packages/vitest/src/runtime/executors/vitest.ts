import { isNodeBuiltin } from 'mlly'
import { normalize } from 'pathe'
import type { ModuleCache, ViteNodeRunnerOptions } from 'vite-node'
import { ViteNodeRunner } from 'vite-node/client'
import { isInternalRequest } from 'vite-node/utils'
import type { ResolvedConfig } from '../../types'
import type { MockMap } from '../../types/mocker'
import { getCurrentEnvironment, getWorkerState } from '../../utils/global'
import { VitestMocker } from '../mocker'
import { rpc } from '../rpc'

export interface ExecuteOptions extends Pick<ViteNodeRunnerOptions, 'moduleCache'> {
  mockMap: MockMap
}

export class VitestExecutor extends ViteNodeRunner {
  public mocker: VitestMocker

  constructor(config: ResolvedConfig, options: ExecuteOptions) {
    super({
      fetchModule: (id) => {
        return rpc().fetch(id)
      },
      resolveId: async (id, importer) => {
        try {
          return await rpc().resolveId(id, importer)
        }
        catch (err) {
          const runtimeError = new Error(err.message)
          // original stack refers to the server part, and doesn't have information where the error happened
          // TODO: stacktrace is broken, example on timers.test.ts
          throw runtimeError
        }
      },
      moduleCache: options.moduleCache,
      interopDefault: config.deps.interopDefault,
      root: config.root,
      base: config.base,
    })

    this.mocker = new VitestMocker(this, options.mockMap)

    Object.defineProperty(globalThis, '__vitest_mocker__', {
      value: this.mocker,
      writable: true,
      configurable: true,
    })
  }

  shouldResolveId(id: string, _importee?: string | undefined): boolean {
    if (isInternalRequest(id))
      return false
    const environment = getCurrentEnvironment()
    // do not try and resolve node builtins in Node
    // import('url') returns Node internal even if 'url' package is installed
    return environment === 'node' ? !isNodeBuiltin(id) : !id.startsWith('node:')
  }

  async resolveUrl(id: string, importee?: string) {
    if (importee && importee.startsWith('mock:'))
      importee = importee.slice(5)
    return super.resolveUrl(id, importee)
  }

  async dependencyRequest(id: string, fsPath: string, callstack: string[]): Promise<any> {
    const mocked = await this.mocker.requestWithMock(fsPath, callstack)

    if (typeof mocked === 'string')
      return super.dependencyRequest(mocked, mocked, callstack)
    if (mocked && typeof mocked === 'object')
      return mocked
    return super.dependencyRequest(id, fsPath, callstack)
  }

  prepareContext(mod: ModuleCache, context: Record<string, any>) {
    const workerState = getWorkerState()

    // support `import.meta.vitest` for test entry
    if (workerState.filepath && normalize(workerState.filepath) === normalize(context.__filename)) {
      // @ts-expect-error injected untyped global
      Object.defineProperty(context.__vite_ssr_import_meta__, 'vitest', { get: () => globalThis.__vitest_index__ })
    }

    return context
  }
}
