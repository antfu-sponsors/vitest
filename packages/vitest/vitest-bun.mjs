#!/usr/bin/env bun
import './dist/cli-wrapper.js'
