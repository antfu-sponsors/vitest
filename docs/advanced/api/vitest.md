---
outline: deep
---

# Vitest

Vitest instance requires the current test mode. It can be either:

- `test` when running runtime tests
- `benchmark` when running benchmarks

## mode

### test

Test mode will only call functions inside `test` or `it`, and throws an error when `bench` is encountered. This mode uses `include` and `exclude` options in the config to find test files.

### benchmark

Benchmark mode calls `bench` functions and throws an error, when it encounters `test` or `it`. This mode uses `benchmark.include` and `benchmark.exclude` options in the config to find benchmark files.

## start

```ts
function start(filters: string[]): Promise<TestRunResult>
```

You can start running tests or benchmarks with `start` method. You can pass an array of strings to filter test files.

## config

The root (or global) config. If workspace feature is enabled, projects will reference this as `globalConfig`.

::: warning
This is Vitest config, it doesn't extend _Vite_ config. It only has resolved values from the `test` property.
:::

## vite

This is a global [`ViteDevServer`](https://vite.dev/guide/api-javascript#vitedevserver).

## state

::: warning
Public `state` is an experimental API. Breaking changes might not follow SemVer, please pin Vitest's version when using it.
:::

Global state stores information about the current tests. It uses the same API from `@vitest/runner` by default, but we recommend using the [Reported API](/advanced/reporters#reported-tasks) instead by calling `state.getReportedEntity()` on the `@vitest/runner` API:

```ts
const task = vitest.state.idMap.get(taskId) // old API
const testCase = vitest.state.getReportedEntity(task) // new API
```

In the future, the old API won't be exposed anymore.

## snapshot

The global snapshot manager. Vitest keeps track of all snapshots using the `snapshot.add` method.

You can get the latest summary of snapshots via the `vitest.snapshot.summay` property.

## cache

Cache manager that stores information about latest test results and test file stats. In Vitest itself this is only used by the default sequencer to sort tests.

## ready

Vitest needs to be resolved with the Vite server to be properly initialized. If the `Vitest` instance was created manually, you might need to check the `ready` status before accessing the `vite`, `state`, `cache`, `config`, and `snapshot` properties; otherwise, they will throw an error in the getter.

In normal circumstances, you would never call this method because `createVitest` and `startVitest` return already resolved Vitest instance.

## getRootTestProject

This returns the root test project. The root project generally doesn't run any tests and is not included in `vitest.projects` unless the user explicitly includes the root config in their workspace.

The primary goal of the root project is to setup the global config. In fact, `rootProject.config` references `rootProject.globalConfig` and `vitest.config` directly.

## provide

Vitest exposes `provide` method which is a shorthand for `vitest.getRootTestProject().provide`. With this method you can pass down values from the main thread to tests. All values are checked with `structuredClone` before they are stored, but the values themselves are not cloned.

To recieve the values in the test, you need to import `inject` method from `vitest` entrypont:

```ts
import { inject } from 'vitest'
const port = inject('wsPort') // 3000
```

For better type safety, we encourage you to augment the type of `ProvidedContext`:

```ts
import { createVitest } from 'vitest/node'

const vitest = await createVitest('test', {
  watch: false,
})
vitest.provide('wsPort', 3000)

declare module 'vitest' {
  export interface ProvidedContext {
    wsPort: number
  }
}
```

::: warning
Technically, `provide` is a method of [`TestProject`](#testproject), so it is limited to the specific project. However, all projects inherit the values from the core project which makes `vitest.provide` universal way of passing down values to tests.
:::

## getProvidedContext

This returns the root context object. This is a shorthand for `vitest.getRootTestProject().getProvidedContext`.

## getProjectByName

This method returns the project by its name. Simillar to calling `vitest.projects.find`.

::: warning
In case the project doesn't exist, this method will return the root project - make sure to check the names again if you need to make sure the project you are looking for is the one returned.
:::

## globTestSpecifications

This method constructs new [test specifications](#testspecification) by collecting every test in all projects with [`project.globTestFiles`](#globtestfiles). It accepts string filters to match the test files.

::: warning
As of Vitest 2.2.0, it's possible to have multiple test specifications with the same module ID (file path) if `poolMatchGlob` has several pools or if `typecheck` is enabled. This possibility will be removed in Vitest 3.
:::

```ts
const specifications = await vitest.globTestSpecifications(['my-filter'])
// [TestSpecification{ moduleId: '/tests/my-filter.test.ts' }]
console.log(specifications)
```

## mergeReports
## collect
## listFiles
## start
## init

## getModuleSpecifications

Returns a list of test specifications related to the module ID. The ID should already be resolved to an absolute file path. If ID doesn't match `include` or `includeSource` patterns, the returned array will be empty.

::: warning
As of Vitest 2.2.0, this method uses a cache to check if the file is a test. To make sure that the cache is not empty, call `globTestSpecifications` at least once.
:::

## runTestSpecifications
## rerunTestSpecifications
## collectTests
## cancelCurrentRun
## updateSnapshot
## invalidateFile
## close
## exit

## shouldKeepServer
## onServerRestart
## onCancel
## onClose
## onTestsRerun
## onFilterWatchedSpecification
