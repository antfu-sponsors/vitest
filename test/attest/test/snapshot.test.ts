import { attest } from '@ark/attest'
import { expect, test } from 'vitest'

test('inline', () => {
  expect(attest(1 + 2)).toMatchInlineSnapshot(`number`)
  expect(attest((1 + 2).toFixed)).toMatchInlineSnapshot(
    `(fractionDigits?: number | undefined) => string`,
  )
})

test('file', () => {
  expect(attest(1 + 2)).toMatchSnapshot()
  expect(attest((1 + 2).toFixed)).toMatchSnapshot()
})

test('mixed', () => {
  expect(1 + 2).toMatchInlineSnapshot(`3`)
  expect(1 + 2).toMatchSnapshot()
  expect(attest(1 + 2)).toMatchInlineSnapshot(`number`)
  expect(attest(1 + 2)).toMatchSnapshot()
})

test('errors', () => {
  expect(attest(1 / 2).type.toString).toMatchInlineSnapshot(`number`)
  expect(
    attest(
      () =>
        // @ts-expect-error test errors
        '1' / 2,
    ).type.errors,
  ).toMatchInlineSnapshot(
    `The left-hand side of an arithmetic operation must be of type 'any', 'number', 'bigint' or an enum type.`,
  )

  // make it work like `.type.toString` by default
  expect(attest(1 / 2)).toMatchInlineSnapshot(`number`)
})

test('completions', () => {
  // TODO(attest) not sure how to do so quick workaround to extract `attest().completions`
  expect({
    $workaroundCompletions: attest(
      () =>
        // @ts-expect-error test completions
        // eslint-disable-next-line dot-notation
        (1 + 2)['to'],
    ),
  }).toMatchInlineSnapshot(`
    {
      "to": [
        "toExponential",
        "toFixed",
        "toLocaleString",
        "toPrecision",
        "toString",
      ],
    }
  `)
})
