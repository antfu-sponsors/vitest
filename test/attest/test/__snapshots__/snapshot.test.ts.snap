// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`file 1`] = `number`;

exports[`file 2`] = `(fractionDigits?: number | undefined) => string`;

exports[`mixed 2`] = `3`;

exports[`mixed 4`] = `number`;
